tests/drivers/clock_control/clock_control_api/src/test_clock_control.c

```diff
@@ -12,99 +12,130 @@ LOG_MODULE_REGISTER(test);
 #include <drivers/clock_control/nrf_clock_control.h>
 #endif
 
+struct device_subsys_data {
+	clock_control_subsys_t subsys;
+	u32_t startup_us;
+};
+
 struct device_data {
 	const char *name;
-	u32_t startup_us;
+	const struct device_subsys_data *subsys_data;
+	u32_t subsys_cnt;
 };
 
 static const struct device_data devices[] = {
 #ifdef DT_INST_0_NORDIC_NRF_CLOCK_LABEL
 	{
-		.name = DT_INST_0_NORDIC_NRF_CLOCK_LABEL  "_16M",
-		.startup_us = IS_ENABLED(CONFIG_SOC_SERIES_NRF91X) ? 3000 : 400
-	},
-
-	{
-		.name = DT_INST_0_NORDIC_NRF_CLOCK_LABEL  "_32K",
-		.startup_us = (CLOCK_CONTROL_NRF_K32SRC == NRF_CLOCK_LFCLK_RC) ?
-				1000 : 300000
+		.name = DT_INST_0_NORDIC_NRF_CLOCK_LABEL,
+		.subsys_data =  (const struct device_subsys_data[]){
+			{
+				.subsys = CLOCK_CONTROL_NRF_SUBSYS_HF,
+				.startup_us =
+					IS_ENABLED(CONFIG_SOC_SERIES_NRF91X) ?
+						3000 : 400
+			},
+			{
+				.subsys = CLOCK_CONTROL_NRF_SUBSYS_LF,
+				.startup_us = (CLOCK_CONTROL_NRF_K32SRC ==
+					NRF_CLOCK_LFCLK_RC) ? 1000 : 300000
+			}
+		},
+		.subsys_cnt = CLOCK_CONTROL_NRF_TYPE_COUNT
 	}
 #endif
 };
 
 
-typedef void (*test_func_t)(const char *dev_name, u32_t startup_us);
+typedef void (*test_func_t)(const char *dev_name,
+			    clock_control_subsys_t subsys,
+			    u32_t startup_us);
 
-typedef bool (*test_capability_check_t)(const char *dev_name);
+typedef bool (*test_capability_check_t)(const char *dev_name,
+					clock_control_subsys_t subsys);
 
-static void setup_instance(const char *dev_name)
+static void setup_instance(const char *dev_name, clock_control_subsys_t subsys)
 {
 	struct device *dev = device_get_binding(dev_name);
 	int err;
 
 	k_busy_wait(1000);
 	do {
-		err = clock_control_off(dev, 0);
+		err = clock_control_off(dev, subsys);
 	} while (err == 0);
 	LOG_INF("setup done");
 }
 
-static void tear_down_instance(const char *dev_name)
+static void tear_down_instance(const char *dev_name,
+				clock_control_subsys_t subsys)
 {
 	struct device *dev = device_get_binding(dev_name);
 
-	clock_control_on(dev, NULL);
+	clock_control_on(dev, subsys);
 }
 
+static void test_with_single_instance(const char *dev_name,
+				      clock_control_subsys_t subsys,
+				      u32_t startup_time,
+				      test_func_t func,
+				      test_capability_check_t capability_check)
+{
+	if ((capability_check == NULL) || capability_check(dev_name, subsys)) {
+		setup_instance(dev_name, subsys);
+		func(dev_name, subsys, startup_time);
+		tear_down_instance(dev_name, subsys);
+		/* Allow logs to be printed. */
+		k_sleep(K_MSEC(100));
+	}
+}
 static void test_all_instances(test_func_t func,
 				test_capability_check_t capability_check)
 {
 	for (int i = 0; i < ARRAY_SIZE(devices); i++) {
```

Comment:
Both `i` and (below) `j` should be declared as either `size_t` or `u32_t` as the upper bound of the loop is an unsigned value.  Zephyr's remarkably bad about this, but it wouldn't hurt to start working to clean it up.

Reply:
done